Email verification

alxdarksage opened the pull request:
- service to add and verify an email address as a verified sender at SES (for consent emails);
- can query to find the status or request to verify the support email address of a study through the API (verification skipped if the address is already verified)
- tied into creating and saving studies. 

I envision using the APIs to give some UI to study developers similar to the AWS console for this, where the status is retrieved and shown separately next to the support email address, and if it is "UNVERIFIED" there will be a button to verify, if pending a button to resend the request.

alxdarksage committed:
Saving

alxdarksage committed:
Tests for the verify email service tests. Service integrated into controller/API and into create/update study workflow.

alxdarksage committed:
Some cleanup looking at the diff

DwayneJengSage commented on app/org/sagebionetworks/bridge/cache/CacheProvider.java:
```diff
@@ -207,7 +209,31 @@ public void removeString(String cacheKey) {
+            promptToStartRedisIfLocal(e);
+            throw new BridgeServiceException(e);
+        }
+    }
```
It looks like email verification calls are fairly cheap. It's probably not worth building cache logic for this.

alxdarksage replied:
Okay, I removed it.

DwayneJengSage commented on test/org/sagebionetworks/bridge/play/controllers/StudyControllerTest.java:
```diff
@@ -107,7 +139,54 @@ public void canSendEmailRoster() throws Exception {
+        controller.getEmailStatus();
+        
+        verify(mockVerificationService).getEmailStatus(EMAIL_ADDRESS);
+    }
```
We should verify the return value, not just that the backing service is called.

alxdarksage replied:
added

DwayneJengSage commented on test/org/sagebionetworks/bridge/services/EmailVerificationServiceTest.java:
```diff
@@ -0,0 +1,125 @@
+        
+        verify(sesClient).getIdentityVerificationAttributes(any());
+        assertEquals(EmailVerificationStatus.VERIFIED, status);
+    }
```
Unit test for null (unverified) status?

alxdarksage replied:
As set up, studies can't be saved without a support email and that's what we pass in. So there's no validation of a submitted value here because no value is submitted from the user. There's emailDoesntExistRequestVerification() which tests what happens if SES has not seen the email before (you get back a null value for value in the attribute map).

alxdarksage replied:
Added these. We check for a couple of nulls that don't happen when testing against SES, plus we have tests for them.

DwayneJengSage commented on app/org/sagebionetworks/bridge/services/EmailVerificationStatus.java:
```diff
@@ -0,0 +1,22 @@
+    public static final EmailVerificationStatus fromSesVerificationStatus(String status) {
+        // From the SES docs, this string can be : "Pending", "Success", "Failed", or "TemporaryFailure". 
+        // We also call Pending == Verified and Sending == Pending. Everything else is unverified 
+        // and is eligible to trigger the verification workflow again.
```
Comment says Pending == Verified and Sending == Pending, which is confusing.

alxdarksage replied:
Updated

alxdarksage committed:
Updated based on comments. Also updated to develop to fix routes.
Merge branch 'develop' of https://github.com/Sage-Bionetworks/BridgePF into email-verification

Conflicts:
	conf/routes

alxdarksage committed:
Do not need "sending"
